db/channel_cache.go

```diff
@@ -214,831 +245,202 @@ func (c *channelCacheImpl) ReleaseLateSequenceClient(channelName string, current
-
-	c.lateLogLock.RLock()
-	latestLog := c._mostRecentLateLog()
-	latestLog.addListener()
-	latestLateSeq = latestLog.logEntry.Sequence
-	c.lateLogLock.RUnlock()
-	return latestLateSeq
-}
-
-// Called when a client (a continuous _changes feed) is no longer referencing the sequence number.
-func (c *singleChannelCache) ReleaseLateSequenceClient(sequence uint64) (success bool) {
-	for _, log := range c.lateLogs {
-		if log.logEntry.Sequence == sequence {
-			log.removeListener()
-			return true
+		if remainingToEvict > 0 {
+			if len(nruEvictionCandidates) > remainingToEvict {
+				evictionElements = append(evictionElements, nruEvictionCandidates[0:remainingToEvict]...)
+				remainingToEvict = 0
+			} else {
+				evictionElements = append(evictionElements, nruEvictionCandidates...)
+				remainingToEvict = remainingToEvict - len(nruEvictionCandidates)
+			}
 		}
-	}
-	return false
-}
-
-// Receive new late sequence
-func (c *singleChannelCache) AddLateSequence(change *LogEntry) {
-	// Add to lateLogs.
-	lateEntry := &lateLogEntry{
-		logEntry:      change,
-		arrived:       time.Now(),
-		listenerCount: 0,
-	}
-	c.lateLogLock.Lock()
-	c.lateLogs = append(c.lateLogs, lateEntry)
-	c.lastLateSequence = change.Sequence
-	// Currently we're only purging on add.  Could also consider a timed purge to handle the case
-	// where all the listeners get caught up, but there aren't any subsequent late entries.  Not
-	// a high priority, as the memory overhead for the late entries should be trivial, and probably
-	// doesn't merit
-	c._purgeLateLogEntries()
-	c.lateLogLock.Unlock()
-}
-
-// Purge entries from the beginning of the list having no active listeners.  Any newly connecting clients
-// will get these entries directly from the cache.  Always maintain
-// at least one entry in the list, to track new listeners.  Expects to have a lock on lateLogLock.
-func (c *singleChannelCache) _purgeLateLogEntries() {
-	for len(c.lateLogs) > 1 && c.lateLogs[0].getListenerCount() == 0 {
-		c.lateLogs = c.lateLogs[1:]
-	}
-}
-
-// Purge entries from the beginning of the list having no active listeners.  Any newly connecting clients
-// will get these entries directly from the cache.  Always maintain
-// at least one entry in the list, to track new listeners.
-func (c *singleChannelCache) purgeLateLogEntries() {
-	c.lateLogLock.Lock()
-	c._purgeLateLogEntries()
-	c.lateLogLock.Unlock()
-}
 
-// mostRecentLateLog assumes caller has at least read lock on c.lateLogLock
-func (c *singleChannelCache) _mostRecentLateLog() *lateLogEntry {
-	if len(c.lateLogs) > 0 {
-		return c.lateLogs[len(c.lateLogs)-1]
-	}
-	return nil
-}
+		cacheSize = c.channelCaches.RemoveElements(evictionElements)
 
-func (c *singleChannelCache) addDocIDs(changes LogEntries) {
-	for _, change := range changes {
-		c.cachedDocIDs[change.DocID] = struct{}{}
+		base.Tracef(base.KeyCache, "Compact iteration complete - eviction count: %d (lwm:%d)", len(evictionElements), c.compactLowWatermark)
```

Comment:
What do you think of adding a couple of new stats to track counts of NRU Evictions and Inactive evictions?

Reply:
Agree - will split that out in https://issues.couchbase.com/browse/CBG-401